pixelflasher: init at 7.11.1.0

cjshearer opened the pull request:
## Description of changes

Add [PixelFlasher](https://github.com/badabing2005/PixelFlasher): "Pixel™ phone flashing GUI utility with features."

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

cjshearer committed:
pixelflasher: init at 7.6.0.0

romner-set commented:
Result of `nixpkgs-review pr 336191` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pixelflasher</li>
  </ul>
</details>

romner-set commented:
Package builds correctly, but crashes on start:
```
Traceback (most recent call last):
  File "PixelFlasher.py", line 35, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "Main.py", line 66, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "advanced_settings.py", line 38, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "runtime.py", line 83, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "payload_dumper.py", line 17, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller/loader/pyimod02_importers.py", line 419, in exec_module
  File "update_metadata_pb2.py", line 33, in <module>
  File "google/protobuf/descriptor.py", line 920, in __new__
TypeError: Descriptors cannot be created directly.
If this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.
If you cannot immediately regenerate your protos, some other possible workarounds are:
 1. Downgrade the protobuf package to 3.20.x or lower.
 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).

More information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates
[2106595] Failed to execute script 'PixelFlasher' due to unhandled exception!
```

cjshearer commented:
@romner-set thanks for the review! I've updated the application and added the workaround `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` to the application wrapper. It now launches for me without error.

maevii reviewed (commented):
Now it seems to work fine, I can't test actual flashing but it starts up and recognizes my pixel correctly.

cjshearer commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,90 @@
+  stdenv,
+  substituteAll,
+}:
+stdenv.mkDerivation (finalAttrs: {
```
I don't think `buildPythonApplication` works with `pyinstaller`, but I'd love to be shown otherwise.

romner-set replied:
I'm not experienced enough with python or packaging it to say either way, so I'm leaving this to the next reviewer/committer. Otherwise everything LGTM, nice work.

samueltardieu commented:
Selecting an image crashes with: `(pixelflasher:128034): GLib-GIO-ERROR **: 09:28:47.941: No GSettings schemas are installed on the system
fish: Job 1, 'pixelflasher' terminated by signal SIGTRAP (Trace or breakpoint trap)`

samueltardieu committed:
pixelflasher: fix crash using gtk3 wrapper

This prevents a crash when using the file
selector.

samueltardieu commented:
@cjshearer You might want to integrate the two latest commits on https://github.com/samueltardieu/nixpkgs/commits/pixelflasher-fix/ into your patch.

Also, if you want, you can add me (`samueltardieu`) as a co-maintainer on this package as I am an heavy user of PixelFlasher. In fact, I was going to submit a package when I saw your PR.

cjshearer committed:
pixelflasher: add samueltardieu as maintainer

cjshearer committed:
pixelflasher: build with python 3.12 and 3.11

cjshearer commented:
@samueltardieu done - thanks!

cjshearer committed:
pixelflasher: mark darwin as broken

cjshearer committed:
pixelflasher: 7.6.0.0 -> 7.9.2.4

cjshearer committed:
pixelflasher: dynamically resolve android-tools path

The default android-tools path was set statically,
which would cause the path to not be resolved
if the android-tools were updated. This commit
fixes the path to be resolved dynamically.

FliegendeWurst commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,101 @@
+    platformdirs
+    protobuf
+    psutil
+    pyinstaller
```
pyinstaller needs to go to `nativeBuildInputs`, as it is used at build time

FliegendeWurst commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,101 @@
+      "platform_tools_path = None" "platform_tools_path = '/run/current-system/sw/bin/'"
+  '';
+
+  buildPhase = ''
+    sh build.sh
+  '';
```
```suggestion
  buildPhase = ''
    runHook preBuild
    chmod +x build.sh
    ./build.sh
    runHook postBuild
  '';
```

FliegendeWurst commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,101 @@
+    wxpython
+  ];
+
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```
Never override patchPhase

FliegendeWurst commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,101 @@
+  '';
+
+  installPhase = ''
+    install -D dist/PixelFlasher $out/bin/pixelflasher
+    install -D images/icon-dark-256.png $out/share/pixmaps/pixelflasher.png
+    ln -s ${finalAttrs.desktopItem}/share/applications $out/share/
```
```suggestion
    runHook preInstall
    install -D dist/PixelFlasher $out/bin/pixelflasher
    install -D images/icon-dark-256.png $out/share/pixmaps/pixelflasher.png
    runHook postInstall
```
Use `copyDesktopItems` instead

FliegendeWurst commented on pkgs/by-name/pi/pixelflasher/package.nix:
```diff
@@ -0,0 +1,101 @@
+    hash = "sha256-bksTk6Zbj8/4FQ7gUGXfJa+kkJE/bFbJF7A4AY6zIRk=";
+  };
+
+  disabled = python3.pythonOlder "3.11";
```
```suggestion
```
Pointless when using stdenv.mkDerivation.

Add `meta.broken = lib.versionOlder python3.version "3.11";` instead.

cjshearer committed:
pixelflasher: fix desktopItems

cjshearer committed:
pixelflasher: mark pyinstaller as build dependency

cjshearer committed:
pixelflasher: mark as broken for python 3.11

cjshearer committed:
pixelflasher: use copyDesktopItems

cjshearer committed:
pixelflasher: patchShebang instead of running with sh

cjshearer committed:
pixelflasher: 7.9.2.4 -> 7.11.1.0

cjshearer commented:
Superseded by #387750
